### DISCUSSION TITLE
Need to update symlink to homebrew-installed python3

### DISCUSSION THREAD

**Discussion by Jmuccigr** [RhetTbull/osxphotos#1092] [Q&A] [answered]
Every so often, osxphotos runs into the problem that the symlink to python3.xx (now 3.11) in ~/.local/pipx/venvs/osxphotos/bin/ is pointing at a non-existent version of that binary, presumably because homebrew has updated it to something more recent. It's easily enough fixed by manually deleting and making a new symlink, but is there any way around either the problem or the manual update?

My python skills are poor. :-(

TIA.

**Comment by RhetTbull** [comment, marked as answer]
This is a homebrew issue. When homebrew updates python, any app installed with `pipx` breaks because the symlink to the old python no longer works.  The easiest thing to do is to run `pipx reinstall-all` which will fix osxphotos and anything else you've installed with pipx.

**Reply by Jmuccigr** [reply]
Thanks. That's a lot easier than me doing it manually.

**Reply by RhetTbull** [reply]
Great! I added a note to the Install section of the README to address this. I wish there was an easier way to distribute python apps but pipx seems to be the easiest given it can update, uninstall, etc. The "homebrew can update the python version pipx is using" thing is a bit of annoyance though.

**Reply by rajscode** [reply]
@RhetTbull Each python3 update that might impact any scheduled crontab or launchctl jobs also for those using osxphotos in those jobs. 

It seems that each time Python3 is updated to a new version (i.e. python 3.9 to 3.11) via brew update / upgrade cycle one may need to re-permission python3 in Privacy and Security. 

My scheduled launchctl job got a Permission Error 1.  I went through the steps manually to repermission python 3.11 in two areas: (a) Files & Folders - Removable Volumes (b) Photos for the Privacy & Security. You'll end up having two instance of python3 showing up there potentially as python3.  More details on this if needed in the #1080 discussion.

**Reply by RhetTbull** [reply]
@rajscode thanks -- that's good to know. Would be nice if homebrew left pipx alone as there's little need to update python for apps that are installed and working. If pipx copied the python to the virtual env instead of symlink that would work too. You could use the pre-built binary included in releases (though I only build for Apple Silicon) but it would have the same problem with permissions on update.

**Reply by rajscode** [reply]
FYI -- re: pipx breaking for osxphotos (and other items) because of updated version of Python being installed by homebrew and the symlink for python3 being repointed to the new version.  The following might help if one wants to do it:  https://github.com/RhetTbull/osxphotos/discussions/1080#discussioncomment-8183742 

NOTE: There are issues with this approach as noted in the comment.

**Reply by RhetTbull** [reply]
FYI, I plan to eventually move to PyApp (ref #1369) to build the osxphotos binary. This is an amazing new packaging tool that can auto-install python, update osxphotos in-place, etc. all while working as a signed macOS binary. It should make packaging much easier than depending on homebrew + pipx which is prone to breakage.
